pkg/consistentfile/file.go

```diff
@@ -0,0 +1,55 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+Package consistentfile provides a mechanism (on Unix-like platforms) to present a consistent view of a file to separate
+processes even while the file is being written.  This is accomplished by writing a temporary file, syncing to disk, and
+renaming over the destination file name.
+
+Partial/inconsistent reads can occur due to:
+ 1. A process attempting to read the file while it is being written to (both in the case of a new file with a
+    short/incomplete write or in the case of an existing, updated file where new bytes may be written at the beginning
+    but old bytes may still be present after).
+ 2. Concurrent goroutines leading to multiple active writers of the same file.
+
+The above mechanism explicitly protects against (1) as all writes are to a file with a temporary name.
+
+There is no explicit protection against multiple, concurrent goroutines attempting to write the same file. However,
+atomically writing the file should mean only one writer will "win" and a consistent file will be visible.
+
+Note: consistentfile is not implemented for Windows. The Windows codepath returns a normal file.
+*/
+package consistentfile
+
+import (
+	"io"
+	"os"
+)
+
+type File interface {
+	io.ReadWriteCloser
+}
+
+// New returns a new consistent file.  On Unix-like platforms, the writer (an io.ReadWriteCloser) is backed by a
+// temporary file placed into the same directory as the destination file (using filepath.Dir to split the directory from
+// the name).  On a call to Close the temporary file is synced to disk and renamed to its final name, hiding any
+// previous file by the same name.
+//
+// Note: Take care to call Close and handle any errors that are returned.  Errors returned from Close may indicate that
+// the file was not written with its final name.
+func New(name string, mode os.FileMode) (File, error) {
```

Comment:
Nice! I guess we no longer need these:
- [WritePidFile](https://github.com/containerd/containerd/blob/2da38b19398b75f194d9fa2c5c07a2e544d8b683/runtime/v2/shim/util.go#L122) - writes a pid file atomically
- [WriteAddress](https://github.com/containerd/containerd/blob/2da38b19398b75f194d9fa2c5c07a2e544d8b683/runtime/v2/shim/util.go#L141) - writes a address file atomically

Reply:
Oh, good catch! I can update to swap those over as well.

Reply:
Done.